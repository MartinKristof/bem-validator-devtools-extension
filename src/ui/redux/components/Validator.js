import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { Col, Container, Row, Table } from 'reactstrap';
import RulesList from './RulesList';
import Heading from './Heading';
import Icon from './icons/Icon';

const Validator = ({ rules, loading, error, isBemDetected, isValid, inspectElement }) => (
  <Container className="text-center">
    <Row className="my-4">
      <Col>
        {!loading && error && (
          <Fragment>
            <Heading>Any error occured!</Heading>
            <Icon kind="error" />
            <p className="secondary">{error}</p>
          </Fragment>
        )}
        {loading && (
          <Fragment>
            <Heading>Loading...</Heading>
            <Icon kind="loader" />
          </Fragment>
        )}
        {!loading && rules.length > 0 && !error && isBemDetected && !isValid && (
          <div>
            <Heading>Invalid!</Heading>
            <div>
              <Icon kind="invalid" />
              <p className="error">BEM errors found!</p>
            </div>
            <Table responsive striped>
              <thead>
                <tr>
                  <td>Missing class</td>
                  <td>Selector</td>
                </tr>
              </thead>
              <tbody>
                <RulesList rules={rules} inspectElement={inspectElement} />
              </tbody>
            </Table>
          </div>
        )}
        {!loading && !error && isBemDetected && isValid && (
          <Fragment>
            <Heading>Valid!</Heading>
            <div>
              <Icon kind="valid" />
              <p className="success">No BEM errors found!</p>
            </div>
          </Fragment>
        )}
        {!loading && !error && !isBemDetected && !isValid && (
          <Fragment>
            <Heading>No BEM classes detected!</Heading>
            <Icon kind="unknown" />
            <p className="secondary">Nothing to lint!</p>
          </Fragment>
        )}
      </Col>
    </Row>
  </Container>
);

Validator.propTypes = {
  rules: PropTypes.arrayOf(
    PropTypes.exact({ missingClassName: PropTypes.string.isRequired, selector: PropTypes.string.isRequired }),
  ).isRequired,
  loading: PropTypes.bool.isRequired,
  error: PropTypes.string,
  isBemDetected: PropTypes.bool.isRequired,
  isValid: PropTypes.bool.isRequired,
  inspectElement: PropTypes.func.isRequired,
};

export default Validator;
