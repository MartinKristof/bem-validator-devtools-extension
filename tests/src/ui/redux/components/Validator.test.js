import React from 'react';
import { shallow } from 'enzyme';
import Validator from '../../../../../src/ui/redux/components/Validator';

describe('<Validator />', () => {
  const inspectElement = jest.fn();

  describe('#render', () => {
    it('should render if rules are valid according BEM', () => {
      const node = shallow(
        <Validator rules={[]} isBemDetected isValid loading={false} inspectElement={inspectElement} error={null} />,
      );

      expect(node).toMatchSnapshot();
    });

    it('should render if rules are invalid according BEM', () => {
      const rules = [{ missingClassName: '.foo', selector: '.bar' }, { missingClassName: '.boo', selector: '.bazz' }];

      const node = shallow(
        <Validator
          rules={rules}
          isBemDetected
          isValid={false}
          loading={false}
          inspectElement={inspectElement}
          error={null}
        />,
      );

      expect(node).toMatchSnapshot();
    });

    it('should render if BEM is not detected', () => {
      const node = shallow(
        <Validator
          rules={[]}
          isBemDetected={false}
          isValid={false}
          loading={false}
          inspectElement={inspectElement}
          error={null}
        />,
      );

      expect(node).toMatchSnapshot();
    });

    it('should render if loading', () => {
      const node = shallow(
        <Validator
          rules={[]}
          isBemDetected={false}
          isValid={false}
          loading
          inspectElement={inspectElement}
          error={null}
        />,
      );

      expect(node).toMatchSnapshot();
    });

    it('should render if error', () => {
      const node = shallow(
        <Validator
          rules={[]}
          isBemDetected={false}
          isValid={false}
          loading={false}
          inspectElement={inspectElement}
          error="Error"
        />,
      );

      expect(node).toMatchSnapshot();
    });
  });
});
