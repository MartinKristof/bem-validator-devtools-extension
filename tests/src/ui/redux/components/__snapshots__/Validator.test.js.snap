// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Validator /> #render should render if BEM is not detected 1`] = `
<Container
  className="text-center"
  tag="div"
>
  <Row
    className="my-4"
    tag="div"
  >
    <Col
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Heading>
        No BEM classes detected!
      </Heading>
      <Icon
        kind="unknown"
        size={64}
      />
      <p
        className="secondary"
      >
        Nothing to lint!
      </p>
    </Col>
  </Row>
</Container>
`;

exports[`<Validator /> #render should render if error 1`] = `
<Container
  className="text-center"
  tag="div"
>
  <Row
    className="my-4"
    tag="div"
  >
    <Col
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Heading>
        Any error occured!
      </Heading>
      <Icon
        kind="error"
        size={64}
      />
      <p
        className="secondary"
      >
        Error
      </p>
    </Col>
  </Row>
</Container>
`;

exports[`<Validator /> #render should render if loading 1`] = `
<Container
  className="text-center"
  tag="div"
>
  <Row
    className="my-4"
    tag="div"
  >
    <Col
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Heading>
        Loading...
      </Heading>
      <Icon
        kind="loader"
        size={64}
      />
    </Col>
  </Row>
</Container>
`;

exports[`<Validator /> #render should render if rules are invalid according BEM 1`] = `
<Container
  className="text-center"
  tag="div"
>
  <Row
    className="my-4"
    tag="div"
  >
    <Col
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <div>
        <Heading>
          Invalid!
        </Heading>
        <div>
          <Icon
            kind="invalid"
            size={64}
          />
          <p
            className="error"
          >
            BEM errors found!
          </p>
        </div>
        <Table
          responsive={true}
          responsiveTag="div"
          striped={true}
          tag="table"
        >
          <thead>
            <tr>
              <td>
                Missing class
              </td>
              <td>
                Selector
              </td>
            </tr>
          </thead>
          <tbody>
            <RulesList
              inspectElement={[MockFunction]}
              rules={
                Array [
                  Object {
                    "missingClassName": ".foo",
                    "selector": ".bar",
                  },
                  Object {
                    "missingClassName": ".boo",
                    "selector": ".bazz",
                  },
                ]
              }
            />
          </tbody>
        </Table>
      </div>
    </Col>
  </Row>
</Container>
`;

exports[`<Validator /> #render should render if rules are valid according BEM 1`] = `
<Container
  className="text-center"
  tag="div"
>
  <Row
    className="my-4"
    tag="div"
  >
    <Col
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Heading>
        Valid!
      </Heading>
      <div>
        <Icon
          kind="valid"
          size={64}
        />
        <p
          className="success"
        >
          No BEM errors found!
        </p>
      </div>
    </Col>
  </Row>
</Container>
`;
